# **Exploratory Data Analisys: Dataset**


This Jupyter notebook is designed as an initial step in an Exploratory Data Analysis (EDA) of a dataset encompassing the fundamentals of companies listed on the Brazilian stock exchange. It aims to provide a structured overview of the financial data, examining the characteristics and distribution of financial indicators through basic descriptive statistics, the identification of unique values, and the detection of outliers. By employing methods such as describe(), unique(), boxplots, correlation matrices, and histograms, the notebook allows users to understand the dataset's scope, including variances and patterns in the data, without venturing into complex analytic theories or models.

The exploration sets the stage for more in-depth analyses in future notebooks by establishing a clear comprehension of the data's initial state and potential data quality issues.

## Initial Setup

### Install Packages

In [2]:
%pip install pandas -q
%pip install plotly -q

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


### Import libs

In [3]:
import os
import pandas as pd
from pathlib import Path
import plotly.express as px
import plotly.graph_objects as go
import plotly.subplots as sp

### Pandas Config

In [4]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Create a file path default

In [5]:
file_path_book = str(Path(os.getcwd()).parent.parent.parent / "data/book")

## Fundamentals Dataset Infos

### Load data

In [6]:
df_fundamentals_book = pd.read_csv(file_path_book + "/fundamentals_book.csv")
df_fundamentals_book.head(5)

ticker                                          long_name  \
0  ABCB4.SA                              Banco ABC Brasil S.A.   
1  AGRO3.SA  BrasilAgro - Companhia Brasileira de Proprieda...   
2  RAIL3.SA                                          Rumo S.A.   
3  ALPA3.SA                                    Alpargatas S.A.   
4  ALPA4.SA                                    Alpargatas S.A.   

               sector                industry    market_cap  enterprise_value  \
0  Financial Services        Banks - Regional  4.265434e+09      1.477339e+10   
1  Consumer Defensive           Farm Products  2.466480e+09      2.912933e+09   
2         Industrials               Railroads  4.228882e+10      5.524305e+10   
3   Consumer Cyclical  Footwear & Accessories  5.309793e+09      6.482982e+09   
4   Consumer Cyclical  Footwear & Accessories  5.350758e+09      6.395236e+09   

   total_revenue  profit_margins  operating_margins  dividend_rate   beta  \
0   1.941779e+09         0.41576            0.38826           1.56  0.679   
1   1.249437e+09         0.21493            0.25031           3.21  0.432   
2   1.031746e+10         0.07639            0.33544           0.07  0.227   
3   4.022153e+09        -0.05671           -0.06434           0.40  0.571   
4   4.022153e+09        -0.05671           -0.06434           0.43  0.571   

         ebitda  trailing_pe  forward_pe     volume  average_volume  \
0  0.000000e+00     4.069768    4.706601    92300.0        747165.0   
1  2.648920e+08     9.450382    6.332481   298100.0        666692.0   
2  4.522541e+09    54.309525   21.723810  5733400.0      14644522.0   
3 -1.980000e+05     0.000000    0.000000     1100.0          3953.0   
4 -1.980000e+05     0.000000   14.555555  1132100.0       5605825.0   

   fifty_two_week_low  fifty_two_week_high  price_to_sales_trailing_12_months  \
0               15.85                21.99                           2.196663   
1               22.29                32.71                           1.974073   
2               16.21                24.44                           4.098764   
3                7.27                17.80                           1.320137   
4                6.81                22.51                           1.330322   

   fifty_day_average  two_hundred_day_average  trailing_annual_dividend_rate  \
0            19.3382                 18.14667                          1.550   
1            27.0106                 25.58635                          3.240   
2            22.5852                 20.95235                          0.066   
3             8.7146                  9.63540                          0.000   
4             8.3228                  9.27290                          0.000   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.080687      24.518       0.785138  7.774306e+09   
1                        0.132029      22.237       1.113460  3.838370e+08   
2                        0.002993       8.334       2.736981  7.656040e+09   
3                        0.000000       7.867       1.008008  4.142880e+08   
4                        0.000000       7.867       0.999110  4.142880e+08   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                35.162  1.829846e+10                      0.001   
1                 3.885  8.720750e+08                      6.801   
2                 4.132  2.184320e+10                      3.935   
3                 0.614  1.550341e+09                      0.000   
4                 0.614  1.550341e+09                      0.000   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0           0.003        0.00000         0.00000           0.01530   
1           0.671        0.25252         0.21201           0.03839   
2           0.121        0.34493         0.43834           0.04252   
3          -0.127        0.43246        -0.00005          -0.00910   
4          -0.127        0.43246   

In [7]:
df_fundamentals_numeric_cols = df_fundamentals_book.select_dtypes(include=["int", "number", "float64"])
df_fundamentals_numeric_cols.head(5)

market_cap  enterprise_value  total_revenue  profit_margins  \
0  4.265434e+09      1.477339e+10   1.941779e+09         0.41576   
1  2.466480e+09      2.912933e+09   1.249437e+09         0.21493   
2  4.228882e+10      5.524305e+10   1.031746e+10         0.07639   
3  5.309793e+09      6.482982e+09   4.022153e+09        -0.05671   
4  5.350758e+09      6.395236e+09   4.022153e+09        -0.05671   

   operating_margins  dividend_rate   beta        ebitda  trailing_pe  \
0            0.38826           1.56  0.679  0.000000e+00     4.069768   
1            0.25031           3.21  0.432  2.648920e+08     9.450382   
2            0.33544           0.07  0.227  4.522541e+09    54.309525   
3           -0.06434           0.40  0.571 -1.980000e+05     0.000000   
4           -0.06434           0.43  0.571 -1.980000e+05     0.000000   

   forward_pe     volume  average_volume  fifty_two_week_low  \
0    4.706601    92300.0        747165.0               15.85   
1    6.332481   298100.0        666692.0               22.29   
2   21.723810  5733400.0      14644522.0               16.21   
3    0.000000     1100.0          3953.0                7.27   
4   14.555555  1132100.0       5605825.0                6.81   

   fifty_two_week_high  price_to_sales_trailing_12_months  fifty_day_average  \
0                21.99                           2.196663            19.3382   
1                32.71                           1.974073            27.0106   
2                24.44                           4.098764            22.5852   
3                17.80                           1.320137             8.7146   
4                22.51                           1.330322             8.3228   

   two_hundred_day_average  trailing_annual_dividend_rate  \
0                 18.14667                          1.550   
1                 25.58635                          3.240   
2                 20.95235                          0.066   
3                  9.63540                          0.000   
4                  9.27290                          0.000   

   trailing_annual_dividend_yield  book_value  price_to_book    total_cash  \
0                        0.080687      24.518       0.785138  7.774306e+09   
1                        0.132029      22.237       1.113460  3.838370e+08   
2                        0.002993       8.334       2.736981  7.656040e+09   
3                        0.000000       7.867       1.008008  4.142880e+08   
4                        0.000000       7.867       0.999110  4.142880e+08   

   total_cash_per_share    total_debt  earnings_quarterly_growth  \
0                35.162  1.829846e+10                      0.001   
1                 3.885  8.720750e+08                      6.801   
2                 4.132  2.184320e+10                      3.935   
3                 0.614  1.550341e+09                      0.000   
4                 0.614  1.550341e+09                      0.000   

   revenue_growth  gross_margins  ebitda_margins  return_on_assets  \
0           0.003        0.00000         0.00000           0.01530   
1           0.671        0.25252         0.21201           0.03839   
2           0.121        0.34493         0.43834           0.04252   
3          -0.127        0.43246        -0.00005          -0.00910   
4          -0.127        0.43246        -0.00005          -0.00910   

   return_on_equity  gross_profits  total_assets_approx  asset_turnover  \
0           0.15680   1.973086e+09         7.774306e+09        0.249769   
1           0.12170   3.155040e+08         3.838370e+08        3.255124   
2           0.05163   3.146360e+09         7.656040e+09        1.347623   
3          -0.04153   1.968303e+09         4.142880e+08        9.708591   
4          -0.04153   1.968303e+09         4.142880e+08        9.708591   

   earnings_growth_rate  dividend_payout_ratio        equity  debt_to_equity  \
0                   0.1          155000.000000 -1.052416e+10       -1.738710   
1             

### Dataset Infos

#### Data Table

In [8]:
df_fundamentals_book.shape

(291, 43)

In [9]:
df_fundamentals_book.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 291 entries, 0 to 290
Data columns (total 43 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   ticker                             291 non-null    object 
 1   long_name                          291 non-null    object 
 2   sector                             291 non-null    object 
 3   industry                           291 non-null    object 
 4   market_cap                         291 non-null    float64
 5   enterprise_value                   291 non-null    float64
 6   total_revenue                      291 non-null    float64
 7   profit_margins                     291 non-null    float64
 8   operating_margins                  291 non-null    float64
 9   dividend_rate                      291 non-null    float64
 10  beta                               291 non-null    float64
 11  ebitda                             291 non-null    float64

In [10]:
df_fundamentals_book.describe()

market_cap  enterprise_value  total_revenue  profit_margins  \
count  2.910000e+02      2.910000e+02   2.910000e+02      291.000000   
mean   1.787795e+10      3.405533e+10   1.721826e+10        0.286119   
std    5.377619e+10      1.138558e+11   5.712366e+10        1.559603   
min    1.257716e+06     -6.435227e+09  -4.242400e+07       -2.162070   
25%    2.799440e+08      6.108281e+08   5.333440e+08        0.006520   
50%    2.451804e+09      4.040345e+09   2.344667e+09        0.080330   
75%    9.835650e+09      1.914586e+10   1.059748e+10        0.207180   
max    4.832119e+11      9.574406e+11   5.815630e+11       21.757490   

       operating_margins  dividend_rate        beta        ebitda  \
count         291.000000     291.000000  291.000000  2.910000e+02   
mean           -0.092924      29.014124    0.744103  4.008850e+09   
std             1.635946     462.561653    0.426138  2.469248e+10   
min           -14.460840       0.000000   -0.190000 -1.011200e+09   
25%             0.038240       0.085000    0.464000  7.221336e+06   
50%             0.143200       0.370000    0.680000  2.535170e+08   
75%             0.239180       1.255000    1.029000  1.805205e+09   
max             1.008160    7891.550000    2.014000  2.910870e+11   

       trailing_pe  forward_pe        volume  average_volume  \
count   291.000000  291.000000  2.910000e+02    2.910000e+02   
mean     11.956751    4.730074  1.397870e+06    3.309691e+06   
std      21.802508    8.039842  6.385571e+06    9.698547e+06   
min       0.000000  -10.142858  0.000000e+00    0.000000e+00   
25%       0.315364    0.000000  2.000000e+02    3.320500e+03   
50%       7.008511    0.000000  2.900000e+04    1.291680e+05   
75%      12.451811    7.758718  6.352000e+05    2.251468e+06   
max     220.238100   68.823524  9.450820e+07    1.063420e+08   

       fifty_two_week_low  fifty_two_week_high  \
count           291.00000           291.000000   
mean             16.10192            49.003587   
std              22.40023           382.659743   
min               0.10000             1.570000   
25%               4.01500             8.940000   
50%               9.08000            16.690000   
75%              20.74000            33.775000   
max             242.05000          6530.000000   

       price_to_sales_trailing_12_months  fifty_day_average  \
count                         291.000000         291.000000   
mean                            3.227017          23.392806   
std                            16.152040          48.424500   
min                            -1.871929           0.669600   
25%                             0.339396           6.006700   
50%                             0.830885          12.508200   
75%                             1.885794          26.488700   
max                           233.914550         711.948400   

       two_hundred_day_average  trailing_annual_dividend_rate  \
count               291.000000                     291.000000   
mean                 29.528437                       1.023897   
std                 157.775895                       3.804884   
min                   1.088800                       0.000000   
25%                   6.008975                       0.000000   
50%                  12.016700                       0.149000   
75%                  25.906325                       0.754000   
max                2670.774700                      57.786000   

       trailing_annual_dividend_yield   book_value  price_to_book  \
count                      291.000000   291.000000     291.000000   
mean                         0.035842    -0.777076       5.480571   
std                          0.053032   365.198328      39.448676   
min                          0.000000 -3962.890000       0.000000   
25%                          0.000000     4.655000       0.579816   
50%                          0.014087     9.920000       0.968048   
75%                          0.049804    21.277000       1.812316   
max 

In [11]:
df_fundamentals_book.nunique()

ticker                               290
long_name                            228
sector                                11
industry                              78
market_cap                           290
enterprise_value                     290
total_revenue                        227
profit_margins                       222
operating_margins                    228
dividend_rate                        137
beta                                 211
ebitda                               210
trailing_pe                          222
forward_pe                           148
volume                               197
average_volume                       283
fifty_two_week_low                   272
fifty_two_week_high                  272
price_to_sales_trailing_12_months    290
fifty_day_average                    290
two_hundred_day_average              290
trailing_annual_dividend_rate        125
trailing_annual_dividend_yield       167
book_value                           228
price_to_book   

#### Correlations

In [12]:
fundamentals_corr = df_fundamentals_numeric_cols
fundamentals_corr.columns
fundamentals_corr = fundamentals_corr.corr()
fundamentals_corr.head()

market_cap  enterprise_value  total_revenue  \
market_cap           1.000000          0.843670       0.821591   
enterprise_value     0.843670          1.000000       0.675644   
total_revenue        0.821591          0.675644       1.000000   
profit_margins      -0.003128         -0.010525      -0.028594   
operating_margins    0.079604          0.069891       0.057417   

                   profit_margins  operating_margins  dividend_rate      beta  \
market_cap              -0.003128           0.079604      -0.018939  0.019453   
enterprise_value        -0.010525           0.069891      -0.016805 -0.000411   
total_revenue           -0.028594           0.057417      -0.016661  0.060641   
profit_margins           1.000000          -0.311960      -0.014690 -0.248073   
operating_margins       -0.311960           1.000000      -0.072953  0.004847   

                     ebitda  trailing_pe  forward_pe    volume  \
market_cap         0.788431    -0.011830    0.088890  0.203687   
enterprise_value   0.531132    -0.031284    0.040925  0.166592   
total_revenue      0.880440    -0.043862    0.067601  0.206989   
profit_margins    -0.005791    -0.011816   -0.064730 -0.029118   
operating_margins  0.041584     0.097417    0.104125  0.033193   

                   average_volume  fifty_two_week_low  fifty_two_week_high  \
market_cap               0.385409            0.088372            -0.011959   
enterprise_value         0.323804            0.066398            -0.011351   
total_revenue            0.361693            0.052057            -0.012243   
profit_margins          -0.033683            0.173645            -0.006242   
operating_margins        0.057819            0.010835            -0.068402   

                   price_to_sales_trailing_12_months  fifty_day_average  \
market_cap                                 -0.015582           0.036404   
enterprise_value                           -0.023906           0.028148   
total_revenue                              -0.040985           0.020596   
profit_margins                              0.391527           0.068329   
operating_margins                          -0.028166          -0.031043   

                   two_hundred_day_average  trailing_annual_dividend_rate  \
market_cap                       -0.003787                       0.150759   
enterprise_value                 -0.004601                       0.114904   
total_revenue                    -0.007690                       0.162399   
profit_margins                    0.010066                       0.029786   
operating_margins                -0.063439                       0.081445   

                   trailing_annual_dividend_yield  book_value  price_to_book  \
market_cap                               0.296205    0.019246      -0.022667   
enterprise_value                         0.224523    0.017487      -0.024948   
total_revenue                            0.334236    0.016996      -0.022205   
profit_margins                          -0.026300    0.053952      -0.029206   
operating_margins                        0.147257   -0.040640       0.014484   

                   total_cash  total_cash_per_share  total_debt  \
market_cap           0.588672             -0.008100    0.631356   
enterprise_value     0.717299              0.010304    0.909718   
total_revenue        0.295011             -0.024416    0.408667   
profit_margins      -0.007916              0.541173   -0.011993   
operating_margins    0.051835             -0.263964    0.055215   

                   earnings_quarterly_growth  revenue_growth  gross_margins  \
market_cap                          0.021919       -0.063475       0.007572   
enterprise_value                    0.007621       -0.059935      -0.080652   
total_revenue                       0.006072       -0.095971      -0.029990   
profit_margins                     -0.003671        0.001915       0.318591   
operating_margins                   0.025746        0.222719       0.214595 

In [13]:
color_scale = [[0, 'rgb(150, 245, 231)'], [0.5, 'rgb(100, 195, 181)'], [1, 'rgb(50, 145, 131)']]
fig = px.imshow(fundamentals_corr, text_auto=True, aspect="auto", width=3000, height=1000, template="plotly_dark", title="Mapa de Calor: Correlação entre as variaveis", color_continuous_scale=color_scale)
fig.show()

##### Dispertion Matrix

In [14]:
fig = px.scatter_matrix(fundamentals_corr, dimensions=fundamentals_corr.columns, title='Gráficos de Dispersão para Todas as Colunas', template="plotly_dark",)
fig.update_traces(marker=dict(color='rgb(100, 195, 181)'))

fig.update_layout(width=4000, height=8000, grid=dict(xgap=0.1, ygap=0.1))

fig.show()

##### Strong Positive Correlations (above 0.7)

- **enterprise_value and total_debt (0.909718)**: Suggesting that companies with higher enterprise values tend to have higher levels of total debt, possibly reflecting financial leverage strategies.

- **dividend_rate and fifty_two_week_high (0.996772)**: Indicates an almost perfect positive correlation, where dividend rates are higher when the stock price is at the peak of its 52-week range.

- **average_volume and volume (0.902085)**: Shows that average trading volume is strongly related to total volume, suggesting consistency in investor interest.

- **fifty_two_week_high and two_hundred_day_average (0.995925)**: Reveals that the 52-week high price is almost perfectly aligned with the 200-day moving average, reflecting a long-term upward trend.

- **ebitda and total_revenue (0.880440)**: Means that companies with higher total revenues generally report higher EBITDA, which is expected as EBITDA is an indicator of operational profitability.

##### Strong Negative Correlations (below -0.7)

- **equity and total_debt (-0.959828)**: A strong negative correlation suggests that higher equity is associated with lower levels of total debt.

- **equity and enterprise_value (-0.925808)**: Indicates that as shareholder equity increases, the enterprise value may decrease, which can occur in scenarios where the company is reducing debt or distributing cash to shareholders.

##### Moderate Positive Correlations (between 0.3 and 0.7)

- **profit_margins and total_cash_per_share (0.541173)**: A moderate positive correlation suggests that companies with higher profit margins may have more cash available per share, which could be indicative of operational efficiency and financial solidity.

- **book_value and fifty_two_week_high (0.649831)**: This moderate positive correlation may indicate that a higher book value is associated with higher stock prices at the annual peak, potentially reflecting a perception of intrinsic value by investors.

- **asset_turnover and roi (0.662555)**: Implies that companies with higher asset turnover (indicating efficient use of assets to generate revenue) also tend to have a higher return on investment.

##### Moderate Negative Correlations (between -0.3 and -0.7)

- **asset_turnover and roce (-0.766321)**: Shows that a higher asset turnover is moderately negatively correlated with return on capital employed, which may occur if assets are generating sales but not necessarily reflecting profitability on the invested capital.

- **roce and roi (-0.624547)**: Indicates a moderate negative correlation between return on capital employed and return on investment, which may suggest that specific investments or projects funded by the employed capital are not generating proportional returns to the expected level.

##### Strong Positive Correlations

**trailing_annual_dividend_rate vs. trailing_annual_dividend_yield**: Correlation of **0.998640**.

* Reason: Both variables are related to annual dividend metrics, with an increase in the dividend rate leading to a higher dividend yield.

* Effect: The high correlation reflects this strong linear relationship.

**dividend_rate vs. trailing_annual_dividend_yield**: Correlation of **0.994015**.

* Reason: Increases in the dividend rate (dividend_rate) tend to result in a higher dividend yield (trailing_annual_dividend_yield).

* Effect: The positive correlation reflects this linear relationship.

**dividend_rate vs. trailing_annual_dividend_rate**: Correlation of **0.994015**.

* Reason: Both variables are directly related to dividend rates.

* Effect: The strong correlation indicates that an increase in the dividend rate (dividend_rate) is directly associated with an increase in the trailing annual dividend rate.

**fifty_two_week_low vs. fifty_two_week_high**: Correlation of **0.988344**.

* Reason: Both variables represent the extremes of a 52week price range, with an increase in the minimum price affecting the maximum price.

* Effect: The high correlation reflects this significant positive linear relationship.

**volume vs. fifty_two_week_low**: Correlation of **0.983632**.

* Reason: Increases in trading volume can influence the 52week low price.

* Effect: The positive correlation reflects the relationship between an increase in trading volume and an increase in the low price.

**total_revenue vs. gross_profits**: Correlation of **0.982982**.

* Reason: Total revenue (total_revenue) and gross profits are directly related to a company's financial performance.

* Effect: The strong correlation shows that an increase in revenue tends to result in higher gross profits.

**total_revenue vs. ebitda**: Correlation of **0.974159**.

* Reason: A company's total revenue (total_revenue) is linked to its earnings before interest, taxes, depreciation, and amortization (EBITDA).

* Effect: The positive correlation reflects the relationship between an increase in revenue and higher EBITDA.

**fifty_two_week_high vs. fifty_day_average**: Correlation of **0.968563**.

* Reason: The 52week high price and the 50day average price are related to the recent performance of an asset.

* Effect: The positive correlation indicates that an increase in the 50day average price can affect the 52week high price.

**market_cap vs. total_revenue**: Correlation of **0.966357**.

* Reason: A company's market capitalization is related to its financial performance represented by total revenue.

* Effect: Larger companies in terms of market capitalization generally generate more revenue, justifying the positive correlation.

**market_cap vs. enterprise_value**: Correlation of **0.963003**.


 * Reason: A company's market capitalization is related to its enterprise value, considering debt.


 * Effect: The positive correlation indicates that larger companies in terms of market capitalization may also have a higher enterprise value, considering debt.

**fifty_day_average vs. two_hundred_day_average**: Correlation of **0.961965**.


 * Reason: Both moving averages are often used in technical analysis to assess price trends.


 * Effect: The strong correlation suggests that changes in the 50day average can affect the 200day average.

**fifty_two_week_high vs. two_hundred_day_average**: Correlation of **0.950189**.


 * Reason: The 52week high price and the 200day average are significant indicators in technical analysis.


 * Effect: The correlation shows that the performance of the 52week high price is linked to the 200day average.

**ebitda vs. gross_profits**: Correlation of **0.939498**.


 * Reason: Earnings before interest, taxes, depreciation, and amortization (EBITDA) and gross profits are related to a company's profitability.


 * Effect: The positive correlation reflects that an increase in EBITDA generally results in higher gross profits.

**fifty_two_week_low vs. two_hundred_day_average**: Correlation of **0.935803**.


 * Reason: The 52week low price and the 200day average are used in technical analysis to identify support and resistance levels.


 * Effect: The positive correlation indicates that the 52week low price is related to the 200day average.

**fifty_two_week_low vs. fifty_day_average**: Correlation of **0.933126**.


 * Reason: Both variables are used to assess shortterm price trends.


 * Effect: The positive correlation reflects the relationship between the 52week low price and the 50day average price.

**dividend_rate vs. trailing_annual_dividend_rate**: Correlation of **0.932857**.


 * Reason: Both variables are related to dividend rates.


 * Effect: The positive correlation suggests that an increase in the dividend rate (dividend_rate) is related to an increase in the trailing annual dividend rate.

**volume vs. fifty_two_week_high**: Correlation of **0.926501**.


 * Reason: Increased trading volume can influence the 52week high price.


 * Effect: The positive correlation reflects the relationship between an increase in trading volume and an increase in the high price.

**market_cap vs. ebitda**: Correlation of **0.907940**.


 * Reason: A company's market capitalization is related to its earnings before interest, taxes, depreciation, and amortization (EBITDA).


 * Effect: The positive correlation shows that larger companies in terms of market capitalization may also have higher EBITDA.

**total_revenue vs. enterprise_value**: Correlation of **0.881551**.


 * Reason: A company's total revenue (total_revenue) is related to its enterprise value, considering debt.


 * Effect: The positive correlation reflects that companies with higher total revenue may also have a higher enterprise value, considering their financial obligations.

**volume vs. fifty_day_average**: Correlation of **0.852918**.


 * Reason: Increased trading volume can influence the 50day average price.


 * Effect: The positive correlation shows that an increase in trading volume can affect the 50day average price.

**average_volume vs. two_hundred_day_average**: Correlation of **0.808796**.


 * Reason: Average trading volume and the 200day average are used in technical analysis to assess price trends.


 * Effect: The positive correlation suggests that average trading volume is related to the 200day average.

**total_cash vs. gross_profits**: Correlation of **0.790580**.


 * Reason: Total cash and gross profits are related to a company's liquidity and profitability.


 * Effect: The positive correlation reflects that companies with more cash generally have higher gross profits.

**average_volume vs. fifty_day_average**: Correlation of **0.784489**.


 * Reason: Average trading volume and the 50day average price are used in technical analysis.
 * Effect: The positive correlation suggests that average trading volume influences the 50day average price.

**enterprise_value vs. ebitda**: Correlation of **0.775812**.


 * Reason: Enterprise value is related to earnings before interest, taxes, depreciation, and amortization (EBITDA).


 * Effect: The positive correlation indicates that enterprise value is linked to EBITDA.

**average_volume vs. fifty_two_week_low**: Correlation of **0.759226**.


 * Reason: Average trading volume and the 52week low price are used in technical analysis to assess support and resistance.


 * Effect: The positive correlation suggests that average trading volume is related to the 52week low price.

**gross_profits vs. gross_margins**: Correlation of **0.755418**.


 * Reason: Gross profits and gross profit margins are related to a company's operational efficiency.


 * Effect: The positive correlation reflects that an increase in gross profits generally results in higher gross profit margins.

**volume vs. two_hundred_day_average**: Correlation of **0.744661**.


 * Reason: Increased trading volume can influence the 200day average.


 * Effect: The positive correlation suggests that trading volume affects the 200day average.

**total_revenue vs. total_debt**: Correlation of **0.707458**.


 * Reason: A company's total revenue (total_revenue) is related to its total debt.


 * Effect: The positive correlation indicates that companies with higher total revenue may also have more total debt.

##### Strong Negative Correlations

**enterprise_value vs. gross_margins**: Correlation of **0.711904**.

 - Reason: Higher gross margins can lead to lower enterprise values.

 - Effect: The negative correlation indicates that companies with higher gross margins tend to have lower enterprise values.

**profit_margins vs. beta**: Correlation of **0.779440**.

- Reason: Higher profit margins can lead to lower beta values.

- Effect: The negative correlation suggests that companies with higher profit margins tend to have lower beta values.

**operating_margins vs. total_cash_per_share**: Correlation of **0.801788**.

 - Reason: Companies with lower operating margins may have higher total cash per share.

 - Effect: The negative correlation indicates that lower operating margins are associated with higher total cash per share.

**beta vs. profit_margins**: Correlation of **0.779440**.

 - Reason: Lower beta values can be associated with higher profit margins.

 - Effect: The negative correlation suggests that companies with lower beta values tend to have higher profit margins.

**beta vs. fifty_two_week_low**: Correlation of **0.717153**.

 - Reason: Lower beta values are associated with higher 52week low prices.

 - Effect: The negative correlation indicates that companies with lower beta values tend to have higher 52week low prices.

**beta vs. gross_margins**: Correlation of **0.747430**.

 - Reason: Companies with higher gross margins tend to have lower beta values.

 - Effect: The negative correlation suggests that higher gross margins are associated with lower beta values.

**beta vs. return_on_equity**: Correlation of **0.820595**.

 - Reason: Higher return on equity is associated with lower beta values.

 - Effect: The negative correlation indicates that companies with higher return on equity tend to have lower beta values.

**fifty_two_week_low vs. beta**: Correlation of **0.717153**.

 - Reason: Lower beta values are associated with higher 52week low prices.

 - Effect: The negative correlation suggests that companies with lower beta values tend to have higher 52week low prices.

**total_cash vs. gross_margins**: Correlation of **0.767081**.

 - Reason: Companies with lower gross margins tend to have higher total cash.

 - Effect: The negative correlation indicates that lower gross margins are associated with higher total cash.

**total_cash_per_share vs. operating_margins**: Correlation of **0.801788**.

 - Reason: Companies with lower operating margins tend to have higher total cash per share.

 - Effect: The negative correlation indicates that lower operating margins are associated with higher total cash per share.

**total_debt vs. gross_margins**: Correlation of **0.754986**.

 - Reason: Companies with higher gross margins tend to have lower total debt.

 - Effect: The negative correlation suggests that higher gross margins are associated with lower total debt.

**gross_margins vs. enterprise_value**: Correlation of **0.711904**.

 - Reason: Higher gross margins can lead to lower enterprise values.

 - Effect: The negative correlation indicates that companies with higher gross margins tend to have lower enterprise values.

**gross_margins vs. beta**: Correlation of **0.747430**.

 - Reason: Companies with higher gross margins tend to have lower beta values.

 - Effect: The negative correlation suggests that higher gross margins are associated with lower beta values.

**gross_margins vs. total_cash**: Correlation of **0.767081**.

 - Reason: Companies with lower gross margins tend to have higher total cash.

 - Effect: The negative correlation indicates that lower gross margins are associated with higher total cash.

**gross_margins vs. total_debt**: Correlation of **0.754986**.

 - Reason: Companies with higher gross margins tend to have lower total debt.

 - Effect: The negative correlation suggests that higher gross margins are associated with lower total debt.

**return_on_equity vs. beta**: Correlation of **0.820595**.

 - Reason: Higher return on equity is associated with lower beta values.

 - Effect: The negative correlation indicates that companies with higher return on equity tend to have lower beta values.

#### Outliers

In [15]:
df = df_fundamentals_book.select_dtypes(include=["float64"])
fig = sp.make_subplots(rows=5, cols=8, subplot_titles=df.columns, shared_yaxes=True, horizontal_spacing=0.01, vertical_spacing=0.1)

for row in range(1, 6):

    for col in range(1, 9):

        col_name = df.columns[col]

        trace = go.Box(y=df[col_name], name=col_name, marker=dict(color='rgb(100, 195, 181)'))

        fig.add_trace(trace, row=row, col=col)

fig.update_layout(title_text="<b>Boxplot: Outliers", height=800, width=2500, template="plotly_dark")

fig.show()

##### Detailed View of Outliers by Column

**market_cap**:

 - Outliers may indicate companies with market capitalizations significantly larger or smaller than the majority, reflecting the presence of market giants or very small companies.

**enterprise_value**: 

 - Outliers may suggest companies with exceptional enterprise values, possibly due to substantial debt or valuable assets.

**total_revenue**:

 - Discrepant values may indicate revenues significantly above or below the average, representing extraordinary variations in financial performance.

**profit_margins and operating_margins**: 

 - Outliers in these columns may represent companies with exceptionally high or low profit or operating margins, reflecting notable financial efficiency or inefficiency.

**dividend_rate**:

 - Outliers may indicate the presence or absence of dividends, with extremely high or low values reflecting unusual dividend payment strategies.

**beta**:

 - Outliers may point to stocks with exceptionally high or low betas, reflecting notable volatility or stability relative to the market.

**ebitda**:

 - Discrepant values may represent profits before interest, taxes, depreciation, and amortization significantly different from the average, suggesting extraordinary financial events.

**trailing_pe and forward_pe**:

 - Outliers in these columns may indicate stocks with very high or low pricetoearnings ratios, reflecting possible market deviations.

**volume and average_volume**:

 - Discrepant values may represent stocks traded in volumes much above or below the average, reflecting unusual investor interest.

**fifty_two_week_low and fifty_two_week_high**:

 - Outliers in these columns may indicate stocks that have reached notable price extremes in the past 52 weeks, reflecting volatility or exceptional performance.

**price_to_sales_trailing_12_months**:

 - Discrepant values may suggest stocks being notably evaluated in relation to sales, reflecting market distortions.

**fifty_day_average and two_hundred_day_average**:

 - Outliers in these columns may indicate exceptionally high or low moving averages, reflecting unusual price trends.

**trailing_annual_dividend_rate and trailing_annual_dividend_yield**:

 - Outliers may represent companies paying exceptionally high or low dividends compared to the market average.

**book_value**:

 - Discrepant values may suggest companies with exceptional book values, reflecting the presence of significant assets or high liabilities.

**price_to_book**:

 - Outliers may indicate stocks that are notably evaluated in relation to book value.

**total_cash and total_cash_per_share**:

 - Outliers may represent companies with large cash reserves or substantially different cash values compared to the average.

**total_debt**:

 - Discrepant values may suggest substantial debt or the absence of debt, reflecting different financial strategies.

**earnings_quarterly_growth and revenue_growth**:

 - Outliers may indicate exceptional earnings or revenue growth rates, suggesting remarkable performance.

**gross_margins and ebitda_margins**:

 - Discrepant values in these columns may represent significantly high or low gross or EBITDA margins, reflecting operational efficiency or notable inefficiency.

**return_on_assets and return_on_equity**:

 - Outliers may indicate exceptional returns on assets or equity, reflecting remarkable financial performance.

**gross_profits and earnings_growth_rate**:

 - Outliers may represent exceptional gross profits or earnings growth rates.

**dividend_payout_ratio and roi**:

 - Discrepant values in these columns may indicate notable dividend payout policies or returns on investment.

##### Macro View of Outliers in Financial Data

**Company Size:**

 - Outliers in market values, enterprise value, and revenue may reflect large conglomerates in diversified sectors or small startups.

**Financial Efficiency:**

 - Outliers in profit margins and operating margins may indicate exceptionally efficient or inefficient companies in various sectors.

**Dividend Policies:**

 - Outliers in dividend rates may represent distinct dividend payment strategies among companies in different sectors.

**Market Volatility:**

 - Exceptional betas may suggest the unique volatility of stocks in specific sectors.

**Financial Events:**

 - Outliers in metrics such as EBITDA may reflect nonrecurring financial events, such as mergers, acquisitions, or restructurings.

**Investor Interest:**

 - Stock volume and moving averages may be influenced by investor interest in specific sectors.

**Business Cycles:**

 - Sectors with distinct economic cycles may lead to significant variations in financial performance.

**Debt Policies:**

 - Different sectors have varied approaches to debt, resulting in varying debt values.

**Operational Performance:**

 - Outliers in profit growth metrics, margins, and returns may reflect the unique performance of companies in different sectors.


#### Histograms and Distribution

In [16]:
import plotly.subplots as sp
import plotly.graph_objects as go

df = df_fundamentals_book.select_dtypes(include=["float64"])
fig = sp.make_subplots(rows=5, cols=8, subplot_titles=df.columns, shared_yaxes=True, horizontal_spacing=0.01, vertical_spacing=0.1)

for row in range(1, 6):
    for col in range(1, 9):
        subplot_num = (row - 1) * 8 + col
        if subplot_num <= len(df.columns):
            col_name = df.columns[subplot_num - 1]
            trace = go.Histogram(x=df[col_name], name=col_name, marker=dict(color = 'rgb(100, 195, 181)'), showlegend=False)
            fig.add_trace(trace, row=row, col=col)

fig.update_layout(title_text="Histogram: Distribution", height=800, width=2500, template="plotly_dark")

fig.show()


**market_cap (Capitalização de Mercado):**

- Most companies have a market capitalization between **1.257716e+06** and **4.832232e+10**.

**enterprise_value (Valor Empresarial):**

- Most companies have an enterprise value between **-6.435227e+09** and **8.995235e+10**.

**total_revenue (Receita Total):**

- Most companies have total revenue between **-4.242400e+07** and **5.811812e+10**.

**profit_margins (Margens de Lucro):**

- Most companies have profit margins between **-2.162070** and **0.229886**.

**operating_margins (Margens Operacionais):**

- Most companies have operating margins close to **0**.

**dividend_rate (Taxa de Dividendos):**

- Most companies do not pay dividends.

**beta:**

- Most companies have a beta below **0.25**.

**ebitda:**

- Most companies have an EBITDA between **-1.011200e+09** and **2.910870e+11**.

**trailing_pe (P/E Ratio Atual):**

- Most companies have a P/E ratio between **0** and **22**.

**forward_pe (P/E Ratio Futuro):**

- Most companies have a future P/E ratio between **-10.142858** and **68.823524**.

**volume:**

- Most companies have trading volumes below **9450820.0**.

**average_volume:**

- Most companies have average trading volumes below **10634197.9**.

**fifty_two_week_low (Preço Mínimo das Ações em 52 Semanas):**

- Most companies have a 52-week low stock price between **0.100** and **24.295**.

**fifty_two_week_high (Preço Máximo das Ações em 52 Semanas):**

- Most companies have a 52-week high stock price between **1.570** and **6530.000**.

**price_to_sales_trailing_12_months (Price/Sales Ratio em 12 Meses):**

- Most companies have a Price/Sales ratio between **-1.871929** and **21.706719**.

**fifty_day_average (Média de 50 Dias das Ações):**

- - Most companies have a 50-day average stock price below **71.79748**.

**two_hundred_day_average (Média de 200 Dias das Ações):**

- Most companies have a 200-day average stock price below **268.05739**.

**trailing_annual_dividend_rate (Taxa de Dividendos Anual Retroativa):**

- Most companies have a trailing annual dividend rate below **5.7786**.

**trailing_annual_dividend_yield (Yield de Dividendos Anual Retroativo):**

- Most companies have a yield of dividend yield retroativo inferior to **0.026773**.

**book_value (Valor Contábil):**

- Most companies have a book value between **-3962.8900** and **21.3875**.

**price_to_book (Price/Book Ratio):**

- Most companies have a Price/Book ratio below **54.66667**.

**total_cash (Total de Caixa):**

- Most companies have total cash between **6.000** and **3.943251e+10**.

**total_cash_per_share (Total de Caixa por Ação):**

- Most companies have total cash per share below **54.2057**.

**total_debt (Dívida Total):**

- Most companies have total debt between **0** and **8.957765e+11**.

**earnings_quarterly_growth (Crescimento dos Lucros Trimestrais):**

- Most companies have quarterly earnings growth between **-0.9990** and **1.8895**.

**revenue_growth (Crescimento da Receita):**

- Most companies have revenue growth between **-0.9830** and **3.6830**.

**gross_margins (Margens Brutas):**

- Most companies have gross margins close to **0**.

**ebitda_margins (Margens EBITDA):**

- Most companies have EBITDA margins close to **0**.

**return_on_assets (ROA):**

- Most companies have a positive **ROA**.

**return_on_equity (ROE):**

- Most companies have a positive **ROE**.

**gross_profits (Lucros Brutos):**

- Most companies have gross profits between **-4.981650e+08** and **3.296165e+11**.

**earnings_growth_rate (Taxa de Crescimento de Ganhos):**

- Most companies have an earnings growth rate between **-99.90** and **2788.60**.

**dividend_payout_ratio (Taxa de Pagamento de Dividendos):**

- Most companies have a dividend payout ratio close to **0**.

**roi (Return on Investment):**

- Most companies have an **ROI** between **-8.232338** and **34.366094**.
